README
NAME
    Wiki::Toolkit - A toolkit for building Wikis.

DESCRIPTION
    Helps you develop Wikis quickly by taking care of the boring bits
    for you.  You will still need to write some code - this isn't an
    instant Wiki.

    Wiki::Toolkit used to be called CGI::Wiki.

SEE ALSO
    Tom Insam's Wiki::Toolkit::Kwiki is however an instant wiki, running
    on a Wiki::Toolkit backend.

SIGNIFICANT RECENT CHANGES

0.70    ?? March 2006
        Support for moderation on a per-node basis.
        Change to the database schema. You need to run cgi-wiki-setupdb to
          upgrade your database to the new version
		Many more plugin points
        
0.69    23 February 2006
        Change to the database schema. You need to run cgi-wiki-setupdb to
          upgrade your database to the new version

0.49    23 September 2003
        Added "ignore_case" option to ->list_nodes_by_metadata

0.47    29 August 2003
        Added metadata_was and metadata_wasnt options to ->list_recent_changes

0.45    11 August 2003
        Removed dependency on Class::Delegation.

0.43    10 July 2003
        Added metadata_isnt constraint to ->list_recent_changes.

SYNOPSIS

      # Set up a wiki object with an SQLite storage backend, and an
      # inverted index/DB_File search backend.  This store/search
      # combination can be used on systems with no access to an actual
      # database server.

      my $store     = Wiki::Toolkit::Store::SQLite->new(
          dbname => "/home/wiki/store.db" );
      my $indexdb   = Search::InvertedIndex::DB::DB_File_SplitHash->new(
          -map_name  => "/home/wiki/indexes.db",
          -lock_mode => "EX" );
      my $search    = Wiki::Toolkit::Search::SII->new(
          indexdb => $indexdb );

      my $wiki      = Wiki::Toolkit->new( store     => $store,
                                      search    => $search );


MAJOR METHODS

    write_node
          $wiki->write_node($node_name, $content, $checksum);
          $wiki->write_node(
              "Calthorpe Arms",
              "A rather nice pub on Gray's Inn Road",
              $checksum,
              { category => [ "Pub", "Pub Food", "Bloomsbury" ] }
          );

    format
          my $cooked = $wiki->format($raw);

    delete_node
          $wiki->delete_node($node_name);

    list_all_nodes
          my @node_names = $wiki->list_all_nodes;

    list_backlinks
          my @links_to_me = $wiki->list_backlinks($node_name);

    list_nodes_by_metadata
          my @pubs = $wiki->list_nodes_by_metadata(
              metadata_type  => "category",
              metadata_value => "Pub" );

    list_recent_changes
          my @last_week_edits  = $wiki->list_recent_changes( days => 7 );
          my @last_ten_changes = $wiki->list_recent_changes(
                                                  last_n_changes => 10 );

    node_exists
          print "Got wombats" if $wiki->node_exists("Wombats");

    retrieve_node
          my $homepage_content = $wiki->retrieve_node("Home Page");

          my %node_data = $wiki->retrieve_node( $node_name );
          print "Last Modified:    $node_data{last_modified}\n";
          print "Current Version:  $node_data{version}\n";
          print "Current Checksum: $node_data{checksum}\n";
          print "Current Content:  $node_data{content}\n";
          print "Categories:       "
              . join(", ", @{$node_data{metadata}{category}}) . "\n";

    verify_checksum
          my $as_i_left_it = $wiki->verify_checksum( $node_name, $checksum );

    search_nodes
          my @nodes = $search->nodes( "camel" );



Changes
0.86
        Upgrade database schema to version 11 (text columns no longer have a
          default value, for compatibility with MySQL strict mode).  To upgrade
          your existing databases to the new schema, run wiki-toolkit-setupdb.
          Use of schema version 10 and upgrades from this to newer versions
          will continue to be supported for any releases made before
          31 December 2021.
        Remove code for dealing with upgrades to database schema version 10
          (which came in over a decade ago).  PLEASE NOTE that if your schema
          version is older than 10 (see the schema_info table in your database)
          then you MUST upgrade to version 10 by running wiki-toolkit-setupdb
          BEFORE you install this version of Wiki::Toolkit.
        Fix an issue with MySQL locking of nodes with long names.

0.85    16 December 2016
        Make ->rename_node try to ensure a valid new node name (depending on
          formatter parameters).
        Add dependency on CGI which is no longer core (RT#118582)

0.84    23 February 2014
        Add new options to Lucy and Plucene backends: (1) content munging
          before search indexing, (2) total exclusion of certain nodes from
          search indexing.

0.83    22 November 2013
        Fix bug in Lucy backend (if you wrote a node Foo Bar and then one
          called Foo, it would delete Foo Bar from its index).
        Make Lucy backend capable of indexing metadata as well.
        Allow weighting of node title in Lucy search.

0.82    12 October 2013
        Fix uninitialised value warnings in Lucy backend

0.81    25 May 2013
        Fix compatibility with DBD::SQLite 1.83+ (RT#84680)
        Added a backend for searching with Lucy - Wiki::Toolkit::Search::Lucy
        Fix POD errors which are fatal with perl 5.18.

0.80    3 May 2012
        Spelling fix in Wiki::Toolkit::Search::Base
        Fixed deprecated use of foreach ... qw (fixes #53 and #54)
        Fixed bug with metadata_was in recent changes (#41).

0.79    14 August 2010
        Fix Win32 test failure http://www.cpantesters.org/cpan/report/8099980
        Declare minimum required Perl version
        (thanks to Alexandr Ciornii for the patches)

0.78    16 December 2009
        Fix various POD errors and add POD testing (#45)
        More recent changes tests (for #41)
        Fix transaction handling in SQLite backend (#49)

0.77    24 December 2008
        Complete support for store->list_metadata_by_type, which
          returns a list of all the metadata values
        Add store->list_metadata_names, which will tell you all the
          different metadata types (names)
        write_node: return the version of the node that was just committed,
        if successful
        Don't write out a new version of a node if the checksum is the
        same as the one already stored (#43)
        t/400_upgrade.t: skip all tests correctly if no backends
        configured (#44)
        Configure charset correctly when passing in an existing $dbh (#24)

0.76    13 July 2008
        Really add missing prereq of DBI!
        Add testing of database schema upgrades (#32)
        Move to new schema version 10, including some missing indexes
          and support for verified flags. Note that
          the code using this column has not yet been written (#25, #34).
        Changed Wiki::Toolkit::Feed::Listing to use metadata_was instead of
          metadata_wasnt, for efficiency reasons.
        list_dangling_links, list_backlinks: don't return internal links
          that are from a node which does not exist. This works around
          database inconsistencies (introduced by manual deletion of nodes)
          which should be fixed by adding constraints to the database (see #38)

0.75	11 May 2008
        Added support for selecting versions by moderation status (see
          OpenGuides ticket #142).
        Added "new_only" parameter to ->list_recent_changes, so you can ask
          for e.g. all pages added in the last week.  This involved a bit of
          a rewrite of ->list_recent_changes, so please keep an eye out for
          any bugs that may have been introduced (I'm not convinced the test
          suite covered everything it should have).
        Fix some badly-formatted POD documentation which accidentally included
          some code
        Fix uninitialized warnings in feed tests
        Fix database setup to ignore tables that aren't Wiki::Toolkit tables
        Fix test suite's incorrect passing of DBIxFTSMySQL connection
          parameters
        Add missing prereq of DBI (caught by the CPAN testers) and don't
          use DBI in Makefile.PL
        Add new scripts wiki-toolkit-delete-node and
          wiki-toolkit-revert-to-date

0.74    9  June 2007
        Added list_last_version_before method, to get the version of all the
         nodes at a given point in time
        Updated version on Text::WikiFormat dependency to avoid long-standing
         test failure with old version
        Bump version number of Wiki::Toolkit::Plugin as this was not done
         the last time it was changed (for CPAN.pm compatibility)

0.73    12 December 2006
        Added parse_feed_timestamp method
        Added encoding parameter to the Feed methods, to allow overriding
          the store's encoding.

0.72    15 September 2006
        Install wiki-toolkit-rename-node

0.71    30 August 2006
        More feed code refactoring, to allow "mini" feeds with just
          name, and optionally distance and location. Typically used in
          search results, especially with Wiki::Toolkit::Plugin::Locator's
        Ability to filter by nodes without certain metadata, by type
          or type+value ("everything without a longitude" or 
          "everything not category=Pub")
        Fix bug in Listing.pm pod output

0.70    6 June 2006
        Feed code has been refactored.

Makefile.PL
use strict;
use 5.006; #by perlver
use ExtUtils::MakeMaker;
use Data::Dumper;

# Large chunks of this were inspired by the Makefile.PL supplied with
# DBIx::FullTextSearch -- thanks!

# See if we already have some config variables set.
use lib "lib";
eval "use Wiki::Toolkit::TestConfig";

# Even if we do have a previous configuration saved, we can over-ride and
# be asked all the questions again by specifying the -s flag or setting
# the appropriate environment variable.
if ($Wiki::Toolkit::TestConfig::configured
    and not (@ARGV and $ARGV[0] eq '-s')
    and not $ENV{WIKI_TOOLKIT_RERUN_CONFIG}) {
    print "\nFor the test suite, we use the database and user info\n"
        . "specified during the previous run.  If you want to change\n"
        . "some or all of the values, run 'perl Makefile.PL -s'.\n\n"
        . "**** REMEMBER THAT THESE TESTS ARE DESTRUCTIVE.  ****\n"
        . "**** DO NOT RUN THEM ON A DATABASE THAT CONTAINS ****\n"
        . "****       OR EVER WILL CONTAIN LIVE DATA.       ****\n"
        . "****                                             ****\n"
        . "**** THE DATABASE VALUES YOU GAVE PREVIOUSLY ARE ****\n"
        . "****   STORED IN Wiki::Toolkit::TestConfig AND   ****\n"
        . "****   WILL BE USED FOR FUTURE INSTALLS OF THIS  ****\n"
        . "****             AND RELATED MODULES.            ****\n\n";
} else {
    print "\nYou should supply at least one set of options for testing,\n"
        . "preferably relevant to the backend(s) you intend to use live.\n"
        . "Running the tests under every possible backend combination is\n"
        . "recommended.  To enter an undefined value, accept the empty\n"
        . "string or explicitly enter 'undef'.\n\n"
        . "****        THESE TESTS ARE DESTRUCTIVE.         ****\n"
        . "**** DO NOT RUN THEM ON A DATABASE THAT CONTAINS ****\n"
        . "****       OR EVER WILL CONTAIN LIVE DATA.       ****\n"
        . "****                                             ****\n"
        . "****  THE DATABASE VALUES YOU GIVE HERE WILL BE  ****\n"
        . "****   STORED IN Wiki::Toolkit::TestConfig AND   ****\n"
        . "****   WILL BE USED FOR FUTURE INSTALLS OF THIS  ****\n"
        . "****             AND RELATED MODULES.            ****\n\n";

    my %config;

    # Grab information from previous runs.
    if ($Wiki::Toolkit::TestConfig::configured) {
        %config = %Wiki::Toolkit::TestConfig::config;
    }

    # Grab information from environment variables.
    foreach my $store (qw(MySQL Pg)) {
        my $dbname = $ENV{"WIKI_TOOLKIT_".uc($store)."_DBNAME"};
        if ($dbname and $dbname ne "undef") {
            $config{$store}{dbname} = $dbname;
            foreach my $var (qw(dbuser dbpass dbhost)) {
                my $value = $ENV{"WIKI_TOOLKIT_".uc($store)."_".uc($var)};
                if ($value and $value ne "undef") {
                    $config{$store}{$var} = $value;
                } elsif ($value eq "undef") {
                    $config{$store}{$var} = undef;
                }
            }
        } elsif ($dbname eq "undef") {
            $config{$store}{dbname} = undef;
        }
    }

    my $dbixfts = $ENV{WIKI_TOOLKIT_DBIXFTS_MYSQL};
    if ($dbixfts and $dbixfts ne "undef") {
        $config{dbixfts} = 1;
    } elsif ($dbixfts eq "undef") {
        $config{dbixfts} = 0;
    }

    # Finally ask questions; then check the settings work.
    my %driver = ( MySQL => "DBD::mysql",
                   Pg    => "DBD::Pg" );
    foreach my $store_type (qw(MySQL Pg)) {
        # See whether we have the driver installed.
        eval "require " . $driver{$store_type};
        if ($@) {
            print "\n$driver{$store_type} not installed... skipping...\n";
            $config{$store_type}{dbname} = undef;
            next;
        }

        # Prompt for the options.
        my ($dbname, $dbuser, $dbpass, $dbhost);
        my $pad = ' ' x (7-length $store_type);
        $dbname = prompt "\n${pad}Database name for $store_type: ",
                         $config{$store_type}{dbname};
        undef $dbname unless ($dbname and $dbname ne "undef");
        if ($dbname and $dbname ne "undef") {
            $dbuser = prompt "            Database user: ",
                             $config{$store_type}{dbuser};
            undef $dbuser unless ($dbuser and $dbuser ne "undef");
            $dbpass = prompt "        Database password: ",
                             $config{$store_type}{dbpass};
            undef $dbpass unless ($dbpass and $dbpass ne "undef");
            $dbhost = prompt "Database host (if needed): ",
                             $config{$store_type}{dbhost};
            undef $dbhost unless ($dbhost and $dbhost ne "undef");

            $config{$store_type}{dbname} = $dbname;
            $config{$store_type}{dbuser} = $dbuser;
            $config{$store_type}{dbpass} = $dbpass;
            $config{$store_type}{dbhost} = $dbhost;
        } else {
            print "\nNo database name supplied... skipping...\n";
            $config{$store_type}{dbname} = undef;
        }
    }

    print "\n";

    # Copy the config hash to the right namespace.
    %Wiki::Toolkit::TestConfig::config = %config;
}

# If we have a MySQL store configured, we can test the DBIx::FullTextSearch
# search backend.
eval {
    require DBIx::FullTextSearch;
    require Lingua::Stem;
};
my $fts_inst = $@ ? 0 : 1;
if ($Wiki::Toolkit::TestConfig::config{MySQL}{dbname} and $fts_inst) {
    print "You have DBIx::FullTextSearch and Lingua::Stem installed,\n";
    print " and a MySQL store configured... configuring for test...\n\n";
    $Wiki::Toolkit::TestConfig::config{dbixfts} = 1;
} else {
    print "Either DBIx::FullTextSearch or Lingua::Stem not installed,\n";
    print "or no MySQL store configured... so won't test that...\n\n";
    $Wiki::Toolkit::TestConfig::config{dbixfts} = undef;
}

# We can test the SQLite backend without asking questions, if it's installed.
eval { require DBD::SQLite; };
if ($@) {
    print "DBD::SQLite not installed... so won't test that...\n\n";
    $Wiki::Toolkit::TestConfig::config{SQLite} = { dbname => undef };
} else {
    print "You have DBD::SQLite... configuring test SQLite database...\n\n";
    $Wiki::Toolkit::TestConfig::config{SQLite} = { dbname => "t/sqlite-test.db" };
}

# If we have Search::InvertedIndex installed, we can test that without
# asking questions.
eval { require Search::InvertedIndex; };
my $sii_inst = $@ ? 0 : 1;
if ( $sii_inst ) {
    print "You have Search::InvertedIndex installed, so will test the S:II\n";
    print "search backend...\n\n";
    $Wiki::Toolkit::TestConfig::config{search_invertedindex} = 1;
} else {
    print "You do not have Search::InvertedIndex installed; skipping test\n";
    print "of S:II search backend...\n\n";
    $Wiki::Toolkit::TestConfig::config{search_invertedindex} = undef;
}

# If we have Plucene installed, we can test that without asking questions.
eval { require Plucene; require File::Spec::Functions };
my $plucene_inst = $@ ? 0 : 1;
if ( $plucene_inst ) {
    print "You have Plucene installed, so will test with that...\n\n";
    $Wiki::Toolkit::TestConfig::config{plucene} = 1;
} else {
    print "Either Plucene or File::Spec::Functions not installed; skipping test...\n\n";
    $Wiki::Toolkit::TestConfig::config{plucene} = undef;
}

# If we have Lucy installed, we can test that without asking questions.
eval { require Lucy; require File::Path };
my $lucy_inst = $@ ? 0 : 1;
if ( $lucy_inst ) {
    print "You have Lucy installed, so will test with that...\n\n";
    $Wiki::Toolkit::TestConfig::config{lucy} = 1;
} else {
    print "Either Lucy or File::Path not installed; skipping test...\n\n";
    $Wiki::Toolkit::TestConfig::config{lucy} = undef;
}

# Write out the config for next run.
open OUT, ">lib/Wiki/Toolkit/TestConfig.pm"
    or die "Couldn't open lib/Wiki/Toolkit/TestConfig.pm for writing: $!";
# warning - blind copy and paste follows.  FIXME.
print OUT Data::Dumper->new([ \%Wiki::Toolkit::TestConfig::config ],
                            [ '*Wiki::Toolkit::TestConfig::config' ]
                           )->Dump,
          "\$Wiki::Toolkit::TestConfig::configured = 1;\n1;\n";
close OUT;

# Some modules are only prerequisites if we intend to test a
# particular backend.
my %config = %Wiki::Toolkit::TestConfig::config;
my %extras;
if ( $config{MySQL}{dbname} ) {
    $extras{'DBD::mysql'} = 0;
}
if ( $config{Pg}{dbname} ) {
    $extras{'DBD::Pg'} = 0;
}
if ( $config{SQLite}{dbname} ) {
    $extras{'DBD::SQLite'} = '0.25'; # some of my tests fail on earlier ones
}
if ( $config{dbixfts} ) {
    $extras{'DBIx::FullTextSearch'} = '0.71'; # earlier ones buggy
}
if ( $config{plucene} ) {
    $extras{'Plucene'} = '1.19'; # earlier has trouble with delete
}

# Write the Makefile.
WriteMakefile( (MM->can('signature_target') ? (SIGN => 1) : ()),
               NAME         => "Wiki::Toolkit",
               VERSION_FROM => "lib/Wiki/Toolkit.pm",
               LICENSE      => "perl_5",
               PREREQ_PM    => { 'Text::WikiFormat' => '0.78', #earlier's buggy
                                 'HTML::PullParser' => 0,
                                 'Digest::MD5'      => 0,
                                 'Test::More'       => 0,
                                 'Time::Piece'      => 0,
                                 'DBI'              => 0,
                                 'CGI'              => 0,
                                 %extras
                               },
               EXE_FILES => [ "bin/wiki-toolkit-setupdb",
                              "bin/wiki-toolkit-rename-node",
                              "bin/wiki-toolkit-delete-node",
                              "bin/wiki-toolkit-revert-to-date"
                            ],
               clean => { FILES => "Config lib/Wiki/Toolkit/TestConfig.pm "
                                 . "t/sqlite-test.db t/sii-db-file-test.db "
                                 . "t/node.db t/plucene t/lucy"
                        },
               META_MERGE        => {
                    "meta-spec" => { version => 2 },
                    resources => {
                        repository => {
                            type => 'git',
                            url =>  'https://github.com/OpenGuides/Wiki-Toolkit.git',
                            web => 'https://github.com/OpenGuides/Wiki-Toolkit',
                        },
                        bugtracker => {
                            url => 'https://github.com/OpenGuides/Wiki-Toolkit/issues',
                        },
                    },
               },
             );
